include/nbl/system/IAsyncQueueDispatcher.h

```diff
@@ -99,7 +116,12 @@ class IAsyncQueueDispatcher : public IThreadHandler<CRTP, InternalStateType>, pu
         request_t& req = request_pool[cb_begin];
         cb_begin = incAndWrapAround(cb_begin);
 
-        static_cast<CRTP*>(this)->process_request(req, optional_internal_state...);
+        req.lock();
+        if (static_cast<CRTP*>(this)->process_request_predicate(req))
+        {
+            static_cast<CRTP*>(this)->process_request(req, optional_internal_state...);
+        }
+        req.unlock();
```

Comment:
yeah, definitely much better to call them `begin` and `end`

Reply:
also why is the `IThreadHandler::lock` locked throughout your `process_request` !?

You've just killed parallelism! (I cannot add a request while any request is being processed)

This is why you need an atomic counter based threadsafe circular buffer.

Reply:
actuall if your CB was threadsafe (and also handled/notified/slept on overflows), there would be no need for a global lock at all!

Every request could have its own lock just for notifying the condvar! (which in C++20 could just be replaced alltogether with a single atomic_flag)

Reply:
> actuall if your CB was threadsafe (and also handled/notified/slept on overflows), there would be no need for a global lock at all!

I dont know to how to make {increment + wrap around} and thats purpose of this global mutex basically. (yes, i know it shouldnt be locked during request processing though, i have a commen about that in the code). I agree that each request could have its own mutex though (just for cvar waiting purposes + user's safe write access into request struct in case of SkippableQueue for what the spinlock is now responsible for)

Reply:
> I dont know to how to make {increment + wrap around} and thats purpose of this global mutex basically.

Wrap around does not need to be atomic, you can compute the wrap around the out-of-date value returned by the fetch and you get the same value as if trying to do an atomic wraparound increment guarded by a mutex.

Reply:
resolved, no more begin/end